ctags/main/promise.c

```diff
@@ -62,6 +62,21 @@ static void attachPromiseModifier (int promise,
 								   void *data);
 
 
+static bool havePromise (const struct promise *p)
+{
+	for (int i = 0; i < promise_count; i++)
+	{
+		struct promise *q = promises + i;
+		if (p->lang == q->lang &&
+			p->startLine == q->startLine && p->startCharOffset == q->startCharOffset &&
+			p->endLine == q->endLine && p->endCharOffset == q->endCharOffset)
```

Comment:
As mentioned in https://github.com/geany/geany/pull/4303#issuecomment-2972662255, what about just comparing the starts?  Of course it could lead to missing tags if a subsequent promise is longer than the current one, but OTOH it's very likely to trigger the issue, and very unlikely not to be an error.  Isn't it?

```suggestion
			p->startLine == q->startLine && p->startCharOffset == q->startCharOffset)
```

OTOH, at least if we drop the identical ones, it's not gonna be infinite, there's only so many (however large) combinations of offsets in a given input.

Reply:
Agreed. Comparing just the beginning reduces the number of possible combinations that have to be parsed in the worst case scenario.

In addition, I think we could also discard promises whose `start > existing_promise_start && start < existing_promise_end && end > existing_promise_end` because such overlapping promises should never be created legally (of course, the problem may be on the already inserted promise side, hard to tell, but this means there's some underlying problem in the parser which has to be fixed anyway). Am I right?